File: apps/csharp/Algorithm/FixedQueue.cs
namespace Algorithm;

/// <summary>第4章 固定長キュー（リングバッファ）</summary>
public class FixedQueue<T>
{
    public class EmptyException() : Exception("キューは空です");
    public class FullException() : Exception("キューは満杯です");

    private readonly object?[] _que;
    private readonly int _capacity;
    private int _front, _rear, _num;

    public FixedQueue(int capacity) { _capacity = capacity; _que = new object[capacity]; }

    public bool IsEmpty() => _num <= 0;
    public bool IsFull() => _num >= _capacity;
    public int Size() => _num;
    public int GetCapacity() => _capacity;

    public void Enque(T value)
    {
        if (IsFull()) throw new FullException();
        _que[_rear] = value; _rear = (_rear + 1) % _capacity; _num++;
    }

    public T Deque()
    {
        if (IsEmpty()) throw new EmptyException();
        T value = (T)_que[_front]!; _front = (_front + 1) % _capacity; _num--;
        return value;
    }

    public T Peek() { if (IsEmpty()) throw new EmptyException(); return (T)_que[_front]!; }

    public int Find(T value)
    {
        for (int i = 0; i < _num; i++)
        {
            int idx = (i + _front) % _capacity;
            if (_que[idx] is T v && v!.Equals(value)) return i;
        }
        return -1;
    }

    public bool Contains(T value) => Find(value) != -1;

    public int Count(T value)
    {
        int c = 0;
        for (int i = 0; i < _num; i++)
        {
            int idx = (i + _front) % _capacity;
            if (_que[idx] is T v && v!.Equals(value)) c++;
        }
        return c;
    }

    public void Clear() { _front = _rear = _num = 0; }
}

File: apps/java/src/main/java/algorithm/FixedQueue.java
package algorithm;

/** 第4章 固定長キュー（リングバッファ） */
public class FixedQueue<T> {

    public static class EmptyException extends RuntimeException {
        public EmptyException() { super("キューは空です"); }
    }

    public static class FullException extends RuntimeException {
        public FullException() { super("キューは満杯です"); }
    }

    private final Object[] que;
    private final int capacity;
    private int front;
    private int rear;
    private int num;

    public FixedQueue(int capacity) {
        this.capacity = capacity;
        this.que = new Object[capacity];
        this.front = 0;
        this.rear = 0;
        this.num = 0;
    }

    public boolean isEmpty() { return num <= 0; }
    public boolean isFull() { return num >= capacity; }
    public int size() { return num; }
    public int getCapacity() { return capacity; }

    public void enque(T value) {
        if (isFull()) throw new FullException();
        que[rear] = value;
        rear++;
        num++;
        if (rear == capacity) rear = 0;
    }

    @SuppressWarnings("unchecked")
    public T deque() {
        if (isEmpty()) throw new EmptyException();
        T value = (T) que[front];
        front++;
        num--;
        if (front == capacity) front = 0;
        return value;
    }

    @SuppressWarnings("unchecked")
    public T peek() {
        if (isEmpty()) throw new EmptyException();
        return (T) que[front];
    }

    public int find(T value) {
        for (int i = 0; i < num; i++) {
            int idx = (i + front) % capacity;
            if (que[idx].equals(value)) return i;
        }
        return -1;
    }

    public boolean contains(T value) {
        return find(value) != -1;
    }

    public int count(T value) {
        int c = 0;
        for (int i = 0; i < num; i++) {
            int idx = (i + front) % capacity;
            if (que[idx].equals(value)) c++;
        }
        return c;
    }

    public void clear() {
        front = rear = num = 0;
    }
}
